README.md
# Individual-based Model framework

*Contributors: Chiel van Amstel, Eloi Martinez-Rabert, Rebeca Gonzalez-Cabaleiro

Individual-based Models (IbM) are characterized for simulating the microbes in an 
aggregate (granule or biofilm) as a discrete entity with unique traits. Each of the microbe is 
a small point of reaction that shapes its kinetics function of the local environment and 
therefore, its growth is affected by the activity of the microbes in the surroundings. Thus, 
IbM consider the intrinsic interaction between diverse microbes, either a positive 
(mutualism, syntrophism…) or negative (competition, ammensalism…) interaction. Diffusion 
of components across the simulation domain is modelled together with the biotic reactions. 
The substrates and products of this activity are diffusing, and these dynamics are solved 
function the boundary conditions imposed over the limits of the domain. This type of models 
is used to describe microbial growth assuming that cellular division occurs a certain 
microbe’s age, cycle stage or size. Therefore, division or death occurs independently for 
each of the cells and only function of local conditions, capturing the heterogeneity of the aggregate 
and supporting non-linear growth models[^1].
____________________________
**Methods of IbM can be downloaded [here](https://github.com/Computational-Platform-IbM/IbM/raw/main/Documents/Methods.pdf).**<br>
Detailed description of [detachment](https://github.com/Computational-Platform-IbM/IbM/blob/main/Documents/detachment.md) and [dynamic time step](https://github.com/Computational-Platform-IbM/IbM/blob/main/Documents/timestepping.md) is also available.

## Before having fun...

**:warning: To open the links in a new tab: right click on the link + "Open link in new tab". :warning:**

### :information_source: Relevant information for reviewers

<!-- There is no peer-review process at this time. -->

#### · ISME Communications Journal
Delete current Main Excel (see [Instructions to Download and Run IbM](https://github.com/Computational-Platform-IbM/IbM#clipboard-instructions-to-download-and-run-ibm)), copy `CmxNitrospira_template.xlsx` from `planning\Excels\Templates` to `planning\Excels\`, and rename it to `main.xlsx`.
<br>To create the seed-file(s) for the other cases, see [Excel setup](https://github.com/Computational-Platform-IbM/IbM#bar_chart-excel-setup-1) of Suspension version (**Metabolic heterogeneity of Comammox *Nitrospira***).

**Thanks for your help in improving the quality of our manuscript.**

<!-- **Thanks for your help in improving the quality of our manuscript.** -->

### :gear: MATLAB installation

This IbM framework is build up in MATLAB. Therefore, MATLAB must be installed in your computer.
<br>You can use the [free 30-day trial](https://www.mathworks.com/campaigns/products/trials.html?ef_id=CjwKCAjwqJSaBhBUEiwAg5W9p96Y1NtC8BCa4Pw_wm3sswXR27ZkvuHZtWMOMUntOrmDSc1Ib3MGCRoCILQQAvD_BwE:G:s&s_kwcid=AL!8664!3!463011314378!p!!g!!matlab%20downlaod&s_eid=ppc_6588247642&q=matlab%20downlaod&gclid=CjwKCAjwqJSaBhBUEiwAg5W9p96Y1NtC8BCa4Pw_wm3sswXR27ZkvuHZtWMOMUntOrmDSc1Ib3MGCRoCILQQAvD_BwE) or "Sign in"/"Create account" if your university already has a MATLAB campus license (see [Campus-Wide License](https://www.mathworks.com/academia/tah-support-program/eligibility.html) search).

Click [here](https://www.mathworks.com/help/install/) for more information about MATLAB Installation and Licensing.

### :clipboard: Instructions to Download and Run IbM

1. Download .zip code. Last version: `v2.4.2`. [Download code](https://github.com/Computational-Platform-IbM/IbM/archive/refs/tags/v2.4.2.zip).
2. Extract files to a destination (:bulb: recommendation - Desktop).
3. Open MATLAB.
    - For more information about MATLAB Layout and how to change it, click [here](https://www.mathworks.com/help/matlab/matlab_env/change-the-desktop-layout.html).
4. Go to the **Code folder<sup>2</sup>**...
    &#09;<br><sup><sup>2</sup>Code folder: folder with `IbM.m` file. </sup>
    &#09;<br>→ writing `cd newFolder` to *Command Window* (more info about [cd](https://www.mathworks.com/help/matlab/ref/cd.html)).
    &#09;<br>→ or using *Folder Toolbar* - paste the folder name where the code was extracted.
5. Generate the path to the main code writing `addpath(genpath('lib'))` to *Command Window*.
6. Create the seed-file<sup>1</sup>:
<br><sup><sup>1</sup>Seed-file: `.mat` file with all information of simulation. This file is used to execute the code.</sup><br>
    1. Modify main Excel (lib\planning\Excels\main.xlsx) with all parameters. 
    <br><sup>Instructions on how to use main.xlsx in *Information* sheet, and [Granule version](https://github.com/Computational-Platform-IbM/IbM#granule-version) | [Suspension version](https://github.com/Computational-Platform-IbM/IbM#suspension-version). (**Excel setup** section)</sup>
    2. Create seed-file writing `create_mat` to *Command Window*.
    3. Save seed-file writing <br> `save('planning/sim_xxxx.mat','grid','bac','constants','init_params','settings','-v7.3')`</br> to *Command Window* (:bulb: `sim_xxxx.mat`, where xxxx is the simulation number [from 0001 to 9999]).
    4. Remove items from workspace, freeing up system memory - writing `clear all` to *Command Window* (more info about [clear](https://www.mathworks.com/help/matlab/ref/clear.html)).
7. Execute IbM code:
    1. Copy the desired seed-file to Code folder (folder with `IbM.m` file)
    2. Call to `IbM(xxxx)` (:bulb: where xxxx is the chosen simulation number).<br>
    Once simulation has been run, `sim_xxxx.mat` (i.e., seed-file) is moved to the corresponding **Results** folder.
8. Get Data (see [Get Data](https://github.com/Computational-Platform-IbM/IbM#get-data) section) or Visualization of Results (see [Data visualization](https://github.com/Computational-Platform-IbM/IbM#data-visualization) section).

### :card_file_box: Supplementary Information

· The seed-file can be reviewed writing `load('sim_xxxx.mat')` (or `load('planning\sim_xxxx.mat')` if this isn't in **Code folder**) to *Command Window*, or double click to `sim_xxxx.mat` in *Current folder*. All data is on *Workspace*.
>**Seed-file structure** (`sim_xxxx.mat`)<p>
`bac`               - Information of initial position (*x,y*), molar mass, radius, species and which cell is active/inactive.
<br>`constants`     - Summary of all parameters of model (Discretization, Kinetic parameters, Metabolic matrix...).
<br>`grid`          - Information of grid.
<br>`init_params`	- Initial conditions of the system (HRT, concentration in aggregate and bulk liquid).
<br>`settings`      - Settings of the model (discretization, HRT, detachment, pH, model version, parallization).

· During the execution of the `IbM()` , *Warning messages* may be displaied to *Command Window*. In general, *Warning messages* are merely informative about the simulation progress. See list of [Warning messages](https://github.com/Computational-Platform-IbM/IbM#warning-list).
<br>A MATLAB error looks like:
> Error using `function` (line x)<br>*Details of error*

## Granule version

### :bar_chart: Excel setup

#### · Microbial spatial distribution study[^2]
:bulb: You can find an Excel template (`granule_template.xlsx`) in `planning\Excels\Templates`. Remember to copy this Excel to `planning\Excels\`, and rename it to `main.xlsx`.<p>
In order to setup the IbM for this study, the following key settings must be selected on the main Excel (see *Settings* sheet in `main.xlsx`):
- ***Dynamic dT***                      - *TRUE*
- ***Variable HRT***                    - *TRUE*
- ***Initialisation method***           - *granule*
- ***kDist***                           - *1.0*
- ***Inactivation enabled***            - *TRUE*
- ***pH bulk concentration corrected*** - *FALSE*
- ***pH solving included***             - *FALSE*
- ***Speciation included***             - *FALSE*
- ***Structure model***                 - *TRUE*
- ***Structure model type***            - *Neut*, *Comp*, *Comm* or *Copr* → changes in the feeding regime.

:warning: *Settings* sheet is a summary of all seetings in the model. **DO NOT CHANGE THEM IN THIS SHEET**, but in the respective tab.

### :newspaper: Publications

```
E. Martinez-Rabert, C. van Amstel, C. Smith, W. T. Sloang, R. Gonzalez-Cabaleiro (2022), 
"Environmental and ecological controls of the spatial distribution of microbial populations in aggregates". 
PLOS Computational Biology. doi: 10.1371/journal.pcbi.1010807
```

## Suspension version

### :bar_chart: Excel setup

#### · Metabolic heterogeneity of Comammox *Nitrospira* [^3]
:bulb: You can find an Excel template (`CmxNitrospira_template.xlsx`) in `planning\Excels\Templates`. Remember to copy this Excel to `planning\Excels\`, and rename it to `main.xlsx`.<p>
In order to setup the IbM for this study, the following key settings must be selected on the main Excel (see *Settings* sheet in `main.xlsx`):
- ***Dynamic dT***                      - *TRUE*
- ***Variable HRT***                    - *FALSE*
- ***Initialisation method***           - *suspension*
- ***kDist***                           - *1.50*
- ***Inactivation enabled***            - *TRUE*
- ***pH bulk concentration corrected*** - *FALSE*
- ***pH solving included***             - *FALSE*
- ***Speciation included***             - *FALSE*
- ***Structure model***                 - *FALSE*
- ***Structure model type***            - *"-"*

:warning: *Settings* sheet is a summary of all seetings in the model. **DO NOT CHANGE THEM IN THIS SHEET**, but in the respective tab.

### :newspaper: Publications

```
E. Martinez-Rabert, C. Smith, W. T. Sloang, R. Gonzalez-Cabaleiro (under review), 
"Competitive and substrate limited environments drive metabolic heterogeneity for Comammox Nitrospira". 
ISME Journal. doi: TBA
```

## Get Data
Relevant results are saved every **dt saved** time (specified in Excel, *Discretization* sheet). Results are saved in the corresponding **Result folder** (`\Results\xxxx`, where `xxxx` is the simulation number) in `results1D.mat` file.
>**Results-file structure** (`\xxxx\results1D.mat`)<p>
`bac_saved`               - Number of cells (*nBacs*), position (*x,y*), radius (*radius*), species/metabolism (*species*), state (*active*), actual growth rate (*mu*) [*columns* - cell iD; *rows* - time].
<br>`conc_saved`          - 3D Matrix of concentration profiles on the transverse plane of aggregate for all substrates [*columns* - x position; *rows* - time; *z* - substrate/product]
<br>`pH_saved`            - Matrix of local pH on the transverse plane of aggregate for all substrates [*columns* - x position; *rows* - time].
<br>`reactor_saved`       - Bulk liquid concentration (*bulk_conc*), Hydraulic Retention Time (*HRT*), density of granule (*granule_density*) [*columns* - substrate/products; *rows* - time].

The data that is saved to the **Results-file** can be modified in the `lib\post_processing\save_slice.m` script.

### :clipboard: Instructions to Get Data
1. Go to the **Result folder** of simulation xxxx (`\Results\xxxx\`) through *Current Folder* browser.
2. Load **Result-file** (`results1D.mat`) - double-click on `.mat` file or writing `load('results1D.mat')` on *Command Window*.
3. Saved data to a file (`.txt`,`.dat`,`.csv` or `.xls*`) using `writematrix(input_data,filename,__)` function (more info about [writematrix](https://www.mathworks.com/help/matlab/ref/writematrix.html)). The file is automatically generated in the **Result folder**.
    &#09;<br><sup>:warning: `writematrix` cannot save *struct*-typed data. Then, `input_data` must start with *struct* name (`bac_saved`, `conc_saved`, `pH_saved` or `reactor_saved`) followed by a dot `.` and the name of variable (e.g., `writematrix(bac_saved.nBacs,'test.xlsx')`).</sup>
    &#09;<br><sup>:bulb: Data can be saved even if the simulation is still running.

## Data visualization

### :clipboard: Instructions for Data visualization
1. Go to the **Code folder** (folder with `IbM.m` file).
2. Generate the path to the main code writing `addpath(genpath('lib'))` to *Command Window* (if not already generated).
3. Run Data visualization → `function()` listed below. 

### · Draw aggregate/s (Granule & Suspension version)
Function to plot cells for a given simulation number and time → `plotBacsMatlab(simulation_number, Time)`.<br>
<sub>Inputs: `simulation_number` is the chosen simulation, and `Time` in days.</sub><br>
<br>**To modify colours, open `lib\debugging\plotBacs.m` script. (Lines 15-17)**<br>
:warning: Make sure that *species_per_color* vector corresponds to number of species/metabolisms.
> If the simulation has not finishied yet, copy the seed-file (`simulation_xxxx.mat`) to the corresponding **Result folder** (`\Results\xxxx`).

### · Draw substrate/product profiles (Granule version)
Function to draw 2D substrate/product profiles → `plot2d_S(simulation_number, substrates_number, Time)`.<br>
<sub>Inputs: `simulation_number` is the chosen simulation, `substrate_number` following substrate order of Excel, and `Time` in days.</sub><br>
<br>**To modify colours, open `lib\lib\post_processing.m` script. (Line 6)**<br>

### · Draw spatial distribution of microbial colonies (Granule version)
Function to plot radial distribution (*layered stratification*) and angular distribution (*columned stratification*) of microbial colonies → `plotStratification(simulation_number, finished, Time)`.<br>
<sub>Inputs: `simulation_number` is the chosen simulation, `finished` Simulation finished: 0 or 1 [N/Y], `Time` in days.</sub><br>
<br>Images `.tif` of plots are automaticatly created to **Result folder** (`\Results\xxxx`). 
> Resolution of Spatial distribution plots can be modified → `lib\post_processing\plotStratification.m` script.
 <br>· **Line 59** for radial distribution (*layered stratification*).
 <br>· **Line 107** for angular distribution (*columned stratification*).

### · Ecological analysis (Suspension version)
Currenlty the platform nOEN (n-Order Ecological Network) is in progress:building_construction:. Scripts to perform the pairwise ecological analysis (2OEN) can be downloaded [here](https://github.com/soundslikealloy/nOEN/tree/main/nOEN.v0).
1. Copy `flocbyfloc.m` and `f2OEN.m` scripts to `lib\post_processing` folder.
2. Run `flocbyfloc(simulation_number, substrates_number, Time)`. An Excel file (`RawData_flocbyfloc.xlsx`) is automatically generated with the mass of all species for each floc at time 0 (*t_0* sheet) and at the end of simulation (*t_max* sheet).<br>
<sub>Inputs: `simulation_number` is the chosen simulation, `substrate_number` following substrate order of Excel, and `Time` in days.</sub><br>
3. Drop .xlsx file (name `data_2OEN.xlsx`) with the list of flocs that you want analyze to `lib\post_processing\Data` folder, following the same structure as `RawData_flocbyfloc.xlsx` (*t_0* and *t_max* sheets).
4. Run `f2OEN` function. This will give the corresponding ecological anlaysis with the sample size of each pirewise interaction and the representation of the data in which Kendall's τ has been calculated.

## Warning List

All possible *Warning messages* are listed here:
<br><sub>Format: *WarningTag* (`script.m` source) - Action/Message</sub>
- *DEBUG:noactionRequired* (`lib\calculate_bulk_concentration.m`) - Negative concentration encountered and corrected.
- *DEBUG:noactionRequired* (`lib\detachment\recalculateT.m`) - Detachment speed equals 0, thus infinite time of crossing.
- *dt_diffusion:Limits* (`lib\integTime.m`) - Maximum/Minimum **dt_diffusion** value is reached.
- *dt_bac:Limits* (`lib\integTime.m`) - Maximum/Minimum **dt_bac** value is reached.
- *Diffusion:SlowConvergence* (`lib\integTime.m`) - **dt_diffusion** is increased. 
- *Diffusion:SlowConvergence* (`lib\integTime.m`) - **dt_bac** is increased. 
- *Diffusion:ConvergenceStuck* (`lib\integTime.m`) - **dt_diffusion** is decreased. 
- *Diffusion:UpwardRES* (`lib\integTime.m`) - **dt_diffusion** is decreased. 
- *Diffusion:NegativeConcentration* (`lib\diffusion\diffusionMG.m`) - **dt_diffusion** is reduced.
- *CellDivision:TooFastDivision* (`lib\integTime.m`) - **dt_bac** is decreased. 
- *BulkConcentration:ValueJump* (`lib\integTime.m`) - **dt_bac** is decreased.

:warning: Special *Warning messages*, **POSSIBLE ACTION REQUIRED**: 
- *DEBUG:actionRequired* (`lib\determine_diffusion_region.m`) - Higher simulation domain must be set.
- *DEBUG:actionRequired* (`lib\calculate_bulk_concentration.m`) - Negative bulk concentration encountered after pH control.
- *DEBUG:actionRequired* (`lib\detachment\computeRoot.m`) - ValueError:Should always be above 0...
- *DEBUG:actionRequired* (`lib\detachment\recalculateT.m`) - All neighbours have infinite time of crossing.

>**Legend**
<br>dt_diffusion - step time applied in solving the diffusion-reaction equation 
<br>dt_bac - step time applied in computation ofbateria mass change, division and inactivation

## Contact

**Eloi Martinez-Rabert**. :envelope: 2424069M@student.gla.ac.uk or eloi.mrp@gmail.com

### References

[^1]: Hellweger, F.L. *et al.*, (2016). *Nature Reviews Microbiology*. doi: 10.1038/nrmicro.2016.62<br>
[^2]: Martinez-Rabert, E. *et al.*, (2022). *PLOS Computational Biology*. doi: 10.1371/journal.pcbi.1010807<br>
[^3]: Martinez-Rabert, E. *et al.*, (under review). *ISME Communications Journal*. doi: TBA<br>


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## [2.4.1] - 2022-10-12
### Added
- Proper README file
- Function to plot substrate/product profiles
- Function to plot spatial distribution of microbial colonies
- Function to perform the ecological analysis - 2OEN.
- Unification of Granule and Suspension versions.
### Changed
- Excel templates. Now specific for under review/published studies.
### Fixed
- Bugs in *Structure model*.
### Fixed
- Unnecessary scripts.
- Debugging code.

## [2.4.0] - 2022-03-27
### Added
- Mechanistic detachment implemented
- Function to plot detachment time per grid cell
- Function to plot bacteria with overlay of detachment times
- Excel template for AOB/NOB/CMX/NRMX/AMX simulation
- Excel template for AOB/NOB/AMX simulation
- Setting for suspension initialisation
- Height correction for single aggregate simulations reaction rate in simulation domain
### Changed
- Loading of model settings from Excel completely refactored
- Also reduce dT_bac when multiple round of division are needed per timestep (overcrowding)
- HRT not only changed when bulk concentrations are higher than setpoint, but also when below the setpoint
- Paramters of adaptive step size optimized
- Diffusion region determined based on center of the grid cell
### Fixed
- Only removing outside bacteria due to erosion, no longer erroneous removal of inside bacteria
- Inactivation method now correctly takes into account the growth rate
- Correctly calculating bulk liquid concentrations (gradient was not calculated when HRT was changed)
- Increase maximum possible number of bacteria per gridcell from 255 to 65535 (uint8 to uint16)
### Removed
- Unnecessary diffusion of non-diffusive compounds (e.g. N2 was in concentration & reaction matrix, but does not diffuse)

## [2.3.0] - 2021-11-15
### Added
- Dynamic dT for diffusion
- Dynamic dT_bac for timestepping
- Automatic switch to parallelisation based on the number of bacteria
- Automatic initialisation of the parallel pool with the correct number of workers
- Automatized .mat creator and registration in simulation_log.md
- kDist constant implementation to shoving algorithm
- New setting: structure model and types (Neutralism, competition, commensalism, co-protection)
- New setting: consideration of pH or not
- Documentation for dynamic dT
### Changed
- Calculus of "correction_concentration_steady_state" based on chosen TolAbs by user
- Independent scripts for mu_max/decay and Monod terms calculus
- All reaction matrix calculations are now in one function, which changes behaviour based on settings.parallelized
- Only structure_model: HRT is recalculated when bulk_conc is higher or lower than setpoint


## [2.2.0] - 2021-10-20
### Added
- Plotting function for reaction matrix (2D profile).
- Improved workflow for simulation using a single file (IbM.m).
- Saving of results.
- Functions for analysing the result files.


## [2.1.0] - 2021-10-19
### Added
- Plotting function for concentrations (2D profile).
- Plotting function for convergence to steady state (norm of delta-concentration).
- Parallelisation of reaction matrix calculations.

### Changed
- Updated the Materials & Methods document.

### Fixed
- Minor bug where incorrect diffusion region was taken in the final smoothing step (multigrid method).


## [2.0.0] - 2021-10-18
### Added
- Docstrings for every function and file
- Check pH after `pH_solve` to be in range [1, 14].
- Check for negative concentrations after `pH_solve`.
- Check for non-convergence using consecutive RES values, continue with steady-state if this is detected.
- Variable for the number of diffusion iterations per steady-state check.
- Shoving algorithm in Java.
- Quadtree datastructure for shoving.
- Complete vectorisation of the bacterial activity (division, death, inactivation, etc.).
- Multigrid solver using V-cycles for solving the diffusion.
- Use of convolutions in the multigrid method for potential further speedup (gpu or FFT method).
- Adaptation in standard multigrid method to only calculate diffusion in the diffusion region.
- Clamp the error correction in Newton-Raphson within the valid range, so that it will always converge.
- Reimplemented the focus-region (only solve smaller domain) for significant speed-up.
- Build-in profiling option.
- Debug warnings for negative concentrations, non-convergence, etc.
- Plotting function for bacteria.
- Plotting function for bulk concentration.
- Plotting function for convergence to steady state (RES value).
- Plotting function for diffusion region.
- Plotting function for maximum error over simulation time (max RES value & delta-concentration).
- Plotting function for profiling.

### Changed
- Complete refactorisation of the code for improved readability and ease of development.
- Restructuring the R.mat struct into separate structs/objects with different functionality.
- Folder structure with better organisation.
- Separation of functionality in different files.
- Calculate kinetics only once per species per gridcell, then using the cumulative mass per species to calculate concentration change.
- Check steady-state based on the maximum RES value instead of the average RES value.
- Use different calculation method (convolutions) to determine the RES values.
- Only check the close neighbours in the shoving algorithms instead of all other bacteria.
- Only compute bulk-layer pH once per reaction matrix calculation.
- 
### Removed
- Check in `pH_solve` to see whether root is present in range [1, 14] (will always hold).
- Bracketing after Newton-Raphson jumps outside of valid pH range.

### Fixed
- Removed redundant function calls.
- Correct function call order in model flow.
- Only perform diffusion in diffusion region, not in bulk layer too.
- Only check RES values in the diffusion region. 


## [1.0.0] - 2021-08-28
Direct import of code from RGC with minor bug fixes.

### Fixed
- Bugs regarding reaction matrix, diffusion and bulk concentration.


mat_creator.m
function mat_creator(num_replicates)
    % Creates .mat files & registrate them in "simulation_log.md"
    % > num_replicates: number of replicates for each simulation
    
    %% argument validation
    arguments
        num_replicates (1,1) {mustBeInteger, mustBeInRange(num_replicates, 1, 9999)}
    end
    
    addpath(genpath('lib')); % make every subfolder with functions accessible to the code
    
    %% IbM version
    version = 'v2.3.0';
    
    %% Last simulation number
    fileparts(which('readwrite_md.py'));
    matID = double(py.readwrite_md.readmd());
    
    %% Creation & registration of simulations 
    excels = dir('planning\Excels\*.xlsx');
    if size(excels,1) == 0
        disp('>> Error 404: Excel files not found.')
    else
        for excel = excels'
            for r = 1:num_replicates
                fprintf('\n')
                disp(['>> Excel: ' excel.name ' (Run ' num2str(r) ')'])
                matID = matID + 1;
                [simname, siminfo, simgoal] = mc(excel.name, matID);
                py.readwrite_md.writemd(matID, simname, siminfo, simgoal, version, r);
                clear R bac constants grid init_params settings
            end
        end
    end
end

function [simname, siminfo, simgoal] = mc(filename, sim_number)
    %{
    Mimicing the integTime from before to see if everything is working.
    Manually setting the structs for bac, grid, constants
    %}

    mat_name = sprintf('sim_%04d.mat', sim_number);
    folder_name = 'planning';
    path = ['planning\Excels\' filename];

    R = loadModelXlsx(path);

    warning('on', 'DEBUG:noActionRequired');
    warning('on', 'DEBUG:actionRequired');

    simname = R.info.simname;
    siminfo = R.info.siminfo;
    simgoal = R.info.simgoal;
    
    settings = struct;
    settings.parallelized = false;
    settings.structure_model = R.settings.structure_model;
    if settings.structure_model
        settings.type = R.settings.type; % {'Neut': Neutralism, 'Comp': Competition, 'Comm': Commensalism, 'Copr': Co-protection}
    end
    settings.pHincluded = R.settings.pHincluded;

    %% Manually creating the structs required for integTime (i.e. loadExcel mimic)

    grid = struct;
    grid.dx = R.Sxy.dx;
    grid.dy = R.Sxy.dy;
    grid.nX = R.Sxy.nx; % for now let's not worry about adding 1 gridcell at the beginning and end of the simulation domain...
    grid.nY = R.Sxy.ny;
    grid.blayer_thickness = R.Sxy.T_blayer;

    bac = struct;
    bac.x = R.bac.atrib(:,1);
    bac.y = R.bac.atrib(:,2);
    n = length(bac.x);
    % n = 5000; radius = R.Sxy.dx * 19; % arbitrary %500 bacs => 4, 5000 bacs => 19
    % [bac.x, bac.y] = rand_circle(n, grid.nX/2*grid.dx, grid.nY/2*grid.dy, radius);
    bac.species = randi(3, size(bac.x)); % random for now
    % bac.species = R.bac.atrib(:,5);
    bac.molarMass = R.bac.atrib(1,3) * ones(n, 1);
    bac.radius = R.bac.atrib(1,6) * ones(n, 1);
    bac.active = ones(size(bac.x), 'logical');
    clear n radius;

    % % retain only a handful of bacteria for faster debugging
    % retain = rand(size(bac.x)) < 0.2;
    % bac.x = bac.x(retain);
    % bac.y = bac.y(retain);
    % bac.species = bac.species(retain);
    % bac.molarMass = bac.molarMass(retain);
    % bac.radius = bac.radius(retain);
    % bac.active = bac.active(retain);

    constants = struct;
    constants.constantpH = R.settings.constantpH;
    constants.pHsetpoint = R.pOp.pH;
    constants.T = R.pOp.T;
    constants.isLiquid = strcmp(R.St.Phase(1:5), 'L') | strcmp(R.St.Phase(1:5), 'P');
    constants.Keq = R.pTh.Keq;
    constants.chrM = R.pTh.chrM;
    constants.Dir_k = logical(R.Inf.Dir_k);
    constants.Vr = R.pOp.Vr;                    % still don't know what this variable means, or more: why it is max_nBacs volume? seems arbitrary
    constants.Vg = (grid.dx ^ 3) * 1000; % L
    constants.StNames = R.St.StNames(1:R.St.numStVLiq2);
    constants.speciesNames = R.rm.rNamesX'; %{'B1', 'B2', 'B3'};
    constants.Ks = R.pTh.Ks; %R.pTh.Ks(:, 1:4);
    constants.Ki = R.pTh.Ki; %R.pTh.Ks(:, 5:7); % Ki > 0 only when co-protection is assumed
    constants.react_v = R.pTh.react_v;
    constants.MatrixMet = R.rm.MatrixMet;
    constants.MatrixDecay = R.rm.MatrixDecay_mod;
    constants.influent_concentrations = R.Inf.St;
    constants.pOp.NH3sp = R.pOp.SP; % Setpoint of substrate S in reactor
    constants.constantN = R.flagN;
    constants.kDist = R.bac.kDist;                 % Extra distance between bacteria, when kDist > 1.
    constants.max_granule_radius = R.bac.bac_ymax; % R.bac.bac_ymax;  % C: see excel          % [m] Maximum radius of granule. To compute the detachment of bacteria when bac_norm > r_max
    constants.dT = R.Sxy.dT;
    constants.dT_bac = R.Sxy.dT_bac;
    constants.dT_divide = R.Sxy.dT_Div;
    constants.dT_save = R.Sxy.dT_Print;
    constants.simulation_end = R.Sxy.maxT;
    constants.diffusion_rates = R.kTr.Diffn;
    constants.diffusion_accuracy = 1e-8; % to be tweaked still
    constants.Tol_a = 1e-12; % in [mol/m3], not [mol/L]! (prev. R.kTr.Tolabs)
    constants.pHtolerance = 1e-15;
    constants.steadystate_tolerance = 0.005; % [0, 1] -> relative/absolute tolerance of steady state
    % constants.correction_concentration_steady_state = 1e-4; % [mol/L]
    constants.correction_concentration_steady_state = R.kTr.Tolabs / constants.steadystate_tolerance; % [mol/L]
    constants.RESmethod = 'max'; % {'mean', 'max', 'norm'}
    constants.bac_MW = R.bac.bac_MW;
    constants.bac_rho = R.bac.bac_rho;
    constants.max_nBac = R.bac.bac_nmax; % <TODO: calculate with better method?>
    constants.inactivationEnabled = R.settings.inactivationBac;
    constants.min_bac_mass_grams = R.bac.bac_mmin;
    constants.max_bac_mass_grams = R.bac.bac_mmax;
    constants.bac_max_radius = R.bac.bac_rmax;
    constants.convergence_accuracy = 1e-6; % should be around 1e-6... the maximum difference between absolute maximum RES values of two diffusion cycles before it is said to no longer converge [RES% / 100]
    constants.nDiffusion_per_SScheck = 2;

    Neumann = 0.2;
    constants.dT = min(grid.dx^2./constants.diffusion_rates * Neumann);

    constants.debug.plotBacteria = false;
    constants.debug.plotConvergence = false; %
    constants.debug.plotMaxErrors = false; %
    constants.debug.plotDiffRegion = false;
    constants.debug.plotBulkConcsOverTime = false; %
    constants.debug.plotProfiling = false; %

    init_params = struct;
    init_params.init_bulk_conc = R.Sxy.Sbc_Dir;
    init_params.init_concs = R.St.StVLiq;
    init_params.invHRT = R.pOp.invHRT;

    %% actual call to integTime
    % directory = 'Testing';
    % constants.simulation_end = 24*7*3; % 3 weeks

    constants.dT_backup = 7*24;

    %{
    bac = bacteria_shove(bac, grid, constants); % otherwise bacteria might overlap at start...
    bac = bacteria_shove(bac, grid, constants); % otherwise bacteria might overlap at start...
    bac = bacteria_shove(bac, grid, constants); % otherwise bacteria might overlap at start...
    bac = bacteria_shove(bac, grid, constants); % otherwise bacteria might overlap at start...
    %}
    if constants.debug.plotBacteria
        plotBacs(grid, bac, constants, 0)
    end

    % temp settings of variables: pls remove
    settings.dynamicDT = true;
    constants.dynamicDT.iterThresholdDecrease = 200;
    constants.dynamicDT.iterThresholdIncrease = 40;
    constants.dynamicDT.nIterThresholdIncrease = 3;
    constants.dynamicDT.initRESThresholdDecrease = 20/100;
    constants.dynamicDT.initRESThresholdIncrease = 20/100;
    
    save(fullfile(folder_name, mat_name), 'grid', 'bac', 'constants', 'init_params', 'settings', '-v7.3')
    
end